### ISSUE TITLE
vSphere CPI hung indefinitely on deploy

### ISSUE THREAD

**Issue by pivotal-todd-robbins** [cloudfoundry/bosh-vsphere-cpi-release#276]
**Describe the bug**

BOSH deploy is ran and director is stuck waiting indefinitely after a VM is created. The last CPI operation that we see executed is set_disk_metadata call. The VM that deploy was stuck on had no job packages created and deploy had been stuck creating this VM for 8+ hours.

**To Reproduce**

The problem is sporadic. The problem was not reproducible after restarting director (to clear up lock) and recreating the VM .

**CPI Error Log**
Full task CPI and debugging will be uploaded.

**Expected behavior**
vSphere CPI calls should not hang indefinitely.

**Release Version & Related Info (please complete the following information):**
 - CPI Version : 53.0.11
 - BOSH Director version: 270.11.1
 - Stemcell Name & Version: 621.76
 - vCenter Version
 - NSX(T/V) Version (If using): none

**Additional context**
This problem causes maintenance windows to lapse as it requires long delays in engaging support for resolution.

**Comment by pivotal-todd-robbins** [comment]
[218907-cpi.txt](https://github.com/cloudfoundry/bosh-vsphere-cpi-release/files/5234266/218907-cpi.txt)
[218907-debug.txt](https://github.com/cloudfoundry/bosh-vsphere-cpi-release/files/5234269/218907-debug.txt)

Attached is bosh task <task ID> --debug and bosh task <task ID> --cpi output from the event.

**Comment by mrosecrance** [comment]
Here's another instance although context on it has been lost by now:
https://cloudfoundry.slack.com/archives/C02HPPYQ2/p1573162453274000

**Comment by ktchen14** [comment]
I'm not sure that this is a CPI problem. Note that the hang occurs when BOSH tries to call `set_disk_metadata`, an operation that's a no-op for the vSphere CPI. Here's the preceding `info` call:

```
D, [2020-09-09T19:06:50.825972 #3336] [create_missing_vm(server/914ead3c-866d-4704-a402-aa7b70463f04 (2)/1)] DEBUG -- DirectorJobRunner: [external-cpi] [cpi-904187] request: {"method":"info","arguments":[],"context":{"director_uuid":"67440434-73aa-4721-9a97-aaa80c9c9bb9","request_id":"cpi-904187","datacenters":"<redacted>","default_disk_type":"<redacted>","host":"<redacted>","password":"<redacted>","user":"<redacted>"}} with command: /var/vcap/jobs/vsphere_cpi/bin/cpi
D, [2020-09-09T19:07:04.842269 #3336] [create_missing_vm(server/914ead3c-866d-4704-a402-aa7b70463f04 (2)/1)] DEBUG -- DirectorJobRunner: [external-cpi] [cpi-904187] response: {"result":{"stemcell_formats":["vsphere-ovf","vsphere-ova"]},"error":null,"log":""}, err: /var/vcap/data/packages/vsphere_cpi/d63d55ff913b828da0b4e0ee135eba198c2781de/lib/ruby_vim_sdk/vmomi_support.rb:56: warning: constant ::Fixnum is deprecated /var/vcap/data/packages/vsphere_cpi/d63d55ff913b828da0b4e0ee135eba198c2781de/lib/nsxt/nsxt_client/api/upgrade_node_upgrade_api.rb:145: warning: character class has '-' without escape: /^[a-zA-Z0-9-.]+$/
I, [2020-09-09T19:06:57.541563 #8646]  INFO -- [req_id cpi-904187]: Starting info...
D, [2020-09-09T19:06:57.543319 #8646] DEBUG -- [req_id cpi-904187]: Running method 'RetrieveServiceContent'...
D, [2020-09-09T19:06:57.611252 #8646] DEBUG -- [req_id cpi-904187]: Running method 'Login'...
I, [2020-09-09T19:07:04.811724 #8646]  INFO -- [req_id cpi-904187]: Finished info in 7.27 seconds
D, [2020-09-09T19:07:04.812272 #8646] DEBUG -- [req_id cpi-904187]: Running method 'Logout'...
, exit_status: pid 8646 exit 0
```

This is on the director PID 3336. PID 8646 is the `fork()`ed CPI process. Unrelated, but we should probably get rid of the `Login`/`Logout` for an `info` operation... Anyways, the normal termination of the CPI is displayed as `, exit_status: pid 8646 exit 0`. Here's the corresponding logging from the CPI logs:

```
I, [2020-09-09T19:06:57.541563 #8646]  INFO -- [req_id cpi-904187]: Starting info...
D, [2020-09-09T19:06:57.543319 #8646] DEBUG -- [req_id cpi-904187]: Running method 'RetrieveServiceContent'...
D, [2020-09-09T19:06:57.611252 #8646] DEBUG -- [req_id cpi-904187]: Running method 'Login'...
I, [2020-09-09T19:07:04.811724 #8646]  INFO -- [req_id cpi-904187]: Finished info in 7.27 seconds
D, [2020-09-09T19:07:04.812272 #8646] DEBUG -- [req_id cpi-904187]: Running method 'Logout'...
```

We can relate these two logs based on the request ID `cpi-904187`. The subsequent call from the director for `set_disk_metadata` however:

```
D, [2020-09-09T19:07:04.842935 #3336] [create_missing_vm(server/914ead3c-866d-4704-a402-aa7b70463f04 (2)/1)] DEBUG -- DirectorJobRunner: [external-cpi] [cpi-349449] request: {"method":"set_disk_metadata","arguments":["disk-1c60790a-248e-41b8-a338-ae045c059958",{"director":"p-bosh","deployment":"service-instance_9ce3ce22-f302-4551-bee0-f5daca684d79","instance_id":"914ead3c-866d-4704-a402-aa7b70463f04","instance_index":"2","instance_group":"server","attached_at":"2020-09-09T19:07:04Z"}],"context":{"director_uuid":"67440434-73aa-4721-9a97-aaa80c9c9bb9","request_id":"cpi-349449","datacenters":"<redacted>","default_disk_type":"<redacted>","host":"<redacted>","password":"<redacted>","user":"<redacted>"},"api_version":1} with command: /var/vcap/jobs/vsphere_cpi/bin/cpi
```

Has request ID `cpi-349449` and no corresponding logs from the CPI. In fact, we should at least see the `Starting [method]` logging message since it happens **outside** of the CPI itself here:

https://github.com/cloudfoundry/bosh-cpi-ruby/blob/master/lib/bosh/cpi/cli.rb#L85

Looking at that code:

```ruby
  def initialize(cpi, logs_string_io, result_io)
    @cpi = cpi
    @logs_string_io = logs_string_io
    @result_io = result_io
    @logger = Bosh::Cpi::Logger.new(STDERR)
  end

  ...

  def run(json)
      ...
      @logger.info("Starting #{method}...")

      cpi = @cpi.call(context, cpi_api_version)
      ...
  end
```

And our code:

```ruby
cpi_lambda = lambda do |context, _|
  ...
  $vc_version = VmodlVersionDiscriminant.retrieve_vmodl_version(
    host, Bosh::Clouds::Config.logger)
  require 'cloud/vsphere'
  Bosh::Clouds::VSphere.new(cloud_properties)
end

cli = Bosh::Cpi::Cli.new(cpi_lambda, soap_log, STDOUT)

cli.run(ARGF.read)
```

Regardless of whether we attempt to connect to the vSphere API, that message should've been logged before any CPI specific code was invoked.

My gut feeling here is that the CPI process wasn't successfully `fork()`ed by the director. If the machine is still around, can we take a look at `dmesg`?

**Comment by mrosecrance** [comment]
The logs issue is tied to a point we raised in https://github.com/cloudfoundry/bosh/issues/2276. There won't be logs, since bosh will only save them when the cpi call returns. I don't know if you also log to a separate location or not. It is possible the forking failed but that seemed really unlikely to me.

**Comment by ktchen14** [comment]
@mrosecrance That's... high unfortunate. Without the logs I think we're just grasping at straws here. Here are our timeouts:

```ruby
      @backing_client.send_timeout = 14400 # 4 hours, for stemcell uploads
      @backing_client.receive_timeout = 14400
      @backing_client.connect_timeout = 60
```

I don't think we can move further on debugging this from the CPI side until the director switches the `capture3` to a `popen3` and streams the logs.

**Comment by mrosecrance** [comment]
Alright sounds like this is parked for now given a lack of visibility into what the underlying issue is. Talked it over briefly with ktchen in another thread. For people that run into this in the future, it would be helpful to see that the child cpi process is there with `ps auxf` and if it's killed, it should allow bosh to grab the logs from the cpi and fail the deploy.

**Comment by ktchen14** [comment]
Hmm I'm now thinking that this is the same issue as #280 .

**Comment by mayrstefan** [comment]
@ktchen14 Initially I also thought our CPI is hanging indefinitly. But when we started collection information for Pivotal/VMware support I realized it times out after four hours like you pointed out in your previous comment. From my understanding this 4h timeout should also be present in CPI release 53.0.11. Because this ticket describes a waiting time much longer than the expected timeout I decided to open a dedicated ticket. But I agree with you that both issues might have the same root cause - they did not receive the response from the vcenter API.
